### PR TITLE
Allow to configure routeName and routePattern dynamically

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

I am targeting this branch, because {reason}.

Closes #{put_issue_number_here}.

## Changelog

```markdown
### Added
- Added `AbstractAdmin::generateBaseRoutePattern()`
- Added `AbstractAdmin::generateBaseRouteName()`

### Deprecated
- Added `AbstractAdmin::$baseRoutePattern` property
- Added `AbstractAdmin::$baseRouteName` property
```

### CODE REVIEW

**Comment by jordisala1991** [comment]
Why is not possible to use the configure method of the admin to change dinamically the values of those 2 properties?

File: src/Admin/AbstractAdmin.php [hunk 1]
```diff
@@ -1856,6 +1800,75 @@ protected function configure(): void
...
+                'Please define a default `baseRoutePattern` value for the admin class `%s`',
+                static::class
+            ));
+        }
+
+        if ($isChildAdmin) {
+            return $this->urlize($matches[5]);
+        }
+
+        return $this->cachedBaseRoutePattern = sprintf(
```
Comment (jordisala1991) [inline src/Admin/AbstractAdmin.php]:
```suggestion
        return sprintf(
```

File: src/Admin/AbstractAdmin.php [hunk 2]
```diff
@@ -1856,6 +1800,75 @@ protected function configure(): void
...
+
+        if (!$matches) {
+            throw new \LogicException(sprintf(
+                'Please define a default `baseRoutePattern` value for the admin class `%s`',
+                static::class
+            ));
+        }
+
+        if ($isChildAdmin) {
+            return $this->urlize($matches[5]);
```
Comment (jordisala1991) [inline src/Admin/AbstractAdmin.php]:
```suggestion
            return $this->urlize($matches[5], '-');
```

**Comment by VincentLanglet** [comment]
> Why is not possible to use the configure method of the admin to change dinamically the values of those 2 properties?

Because the value is cached. The issue was reported on the pageBundle topics.

**Comment by eerison** [comment]
Hey @VincentLanglet what is missing in this ticket? can I help somehow?

**Comment by VincentLanglet** [comment]
I need time to fix the test but feel free to do it

**Comment by eerison** [comment]
> I need time to fix the test but feel free to do it

~~if I get some free time tomorrow I'll try :)~~

Edit 1: I won't be able to fix your pipelines :/, I started to check this ticket: https://github.com/sonata-project/SonataPageBundle/issues/1442

**Review by jordisala1991 — APPROVED** [review]

**Comment by jordisala1991** [comment]
Thanks @VincentLanglet

**Pull request merged**